src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,350 @@
+variables (M : Type u2) [add_comm_group M] [module R M]
+
+namespace tensor_algebra
+
+-- This inductive type is used to express representatives
+-- of the tensor algebra.
+@[nolint doc_blame]
+inductive pre
+| of : M → pre
+| of_scalar : R → pre
+| add : pre → pre → pre
+| mul : pre → pre → pre
+
+namespace pre
+
+instance : inhabited (pre R M) := ⟨of_scalar 0⟩
+
+-- These instances are only used to simplify the notation.
+@[nolint doc_blame]
+def has_coe_module : has_coe M (pre R M) := ⟨of⟩
+@[nolint doc_blame]
+def has_coe_ring : has_coe R (pre R M) := ⟨of_scalar⟩
+@[nolint doc_blame]
+def has_mul : has_mul (pre R M) := ⟨mul⟩
+@[nolint doc_blame]
+def has_add : has_add (pre R M) := ⟨add⟩
+@[nolint doc_blame]
+def has_zero : has_zero (pre R M) := ⟨of_scalar 0⟩
+@[nolint doc_blame]
+def has_one : has_one (pre R M) := ⟨of_scalar 1⟩
+@[nolint doc_blame]
+def has_scalar : has_scalar R (pre R M) := ⟨λ r m, mul (of_scalar r) m⟩
+@[nolint doc_blame]
+def has_neg : has_neg (pre R M) := ⟨mul (of_scalar (-1))⟩
+
+end pre
+
+local attribute [instance]
+  pre.has_coe_module pre.has_coe_ring pre.has_mul pre.has_add pre.has_zero
+  pre.has_one pre.has_scalar pre.has_neg
+
+@[nolint doc_blame]
+def lift_fun {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) : pre R M → A :=
+  λ t, pre.rec_on t f (algebra_map _ _) (λ _ _, (+)) (λ _ _, (*))
+
+@[nolint doc_blame]
+inductive rel : (pre R M) → (pre R M) → Prop
+-- force of to be linear
+| add_lin {a b : M} : rel ↑(a+b) (↑a + ↑b)
+| smul_lin {r : R} {a : M} : rel ↑(r • a) (↑r * ↑a)
+-- force const to be a central ring morphism
+| add_scalar {r s : R} : rel ↑(r + s) (↑r + ↑s)
+| mul_scalar {r s : R} : rel ↑(r * s) (↑r * ↑s)
+| central_scalar {r : R} {a : pre R M} : rel (r * a) (a * r)
+-- commutative additive group
+| add_assoc {a b c : pre R M} : rel (a + b + c) (a + (b + c))
+| add_comm {a b : pre R M} : rel (a + b) (b + a)
+| neg_add {a : pre R M} : rel (-a + a) 0
+| zero_add {a : pre R M} : rel (0 + a) a
+-- multiplicative monoid
+| mul_assoc {a b c : pre R M} : rel (a * b * c) (a * (b * c))
+| one_mul {a : pre R M} : rel (1 * a) a
+| mul_one {a : pre R M} : rel (a * 1) a
+-- distributivity
+| left_distrib {a b c : pre R M} : rel (a * (b + c)) (a * b + a * c)
+| right_distrib {a b c : pre R M} : rel ((a + b) * c) (a * c + b * c)
+-- compatibility
+| add_compat_left {a b c : pre R M} : rel a b → rel (a + c) (b + c)
+| add_compat_right {a b c : pre R M} : rel a b → rel (c + a) (c + b)
+| mul_compat_left {a b c : pre R M} : rel a b → rel (a * c) (b * c)
+| mul_compat_right {a b c : pre R M} : rel a b → rel (c * a) (c * b)
+
+end tensor_algebra
+
+/--
+The tensor algebra of the module `M` over the commutative ring `R`.
+-/
+def tensor_algebra := quot (tensor_algebra.rel R M)
+
+namespace tensor_algebra
+
+local attribute [instance]
+  pre.has_coe_module pre.has_coe_ring pre.has_mul pre.has_add pre.has_zero
+  pre.has_one pre.has_scalar pre.has_neg
+
+instance : ring (tensor_algebra R M) :=
```

Comment:
This proof and the following one really seem to call for a small tactic, something like the `pi_instance` tactic.